### PR TITLE
chore: Factor out reusable workflow for integration tests

### PR DESCRIPTION
**Pull request by robin-aws**
Progress on #2010.

This has no effect on the CI for PRs yet (aside from cleaning up the matrix configuration a bit), but introduces a "nightly build" that runs the integration tests on all platforms as well. This will be used in future changes to have the PR CI only run these tests for one platform, and having the nightly build to run the more expensive matrix instead.

Merging this change first has the benefit of allowing the nightly build to run a few times first, so that the subsequent change can add the automation to check for a successful nightly build before allowing a PR to be merged or a release to be published.

### CODE REVIEW

**Comment by robin-aws** [comment]
Note that since this changes the names of the required checks, I'll have to edit our configuration to match. I'll wait until this is approved and ready to go first before I do that though.

File: .github/workflows/deep-tests.yml
Comment (atomb) [inline .github/workflows/deep-tests.yml:16]: I didn't know you could do this in GitHub workflows! Very nice.

File: .github/workflows/integration-tests-reusable.yml
Comment (atomb) [inline .github/workflows/integration-tests-reusable.yml:52]: Tiny nit, but the rest of the code has spaces after `{{` and before `}}`.

**Review by fabiomadge — DISMISSED** [review]

File: .github/workflows/integration-tests-reusable.yml
Comment (atomb) [inline .github/workflows/integration-tests-reusable.yml:89]: So do Windows jobs run reliably now? That'd be a nice improvement.

**Review by atomb — DISMISSED** [review]

File: .github/workflows/integration-tests-reusable.yml
Comment (robin-aws) [reply .github/workflows/integration-tests-reusable.yml:52]: That's fair, I'm going to batch that up with the next PR so I don't have to wait on the CI again if that's already.

Comment (atomb) [reply .github/workflows/integration-tests-reusable.yml:52]: Yeah, not important at all!

**Review by atomb — APPROVED** [review]

**Pull request merged**
